SWIFT_PACKAGE copt not correctly applied when importing Objective-C library

When a Swift target depends on an Objective-C target that uses `#if SWIFT_PACKAGE` we are currently only adding the `-DSWIFT_PACKAGE` to `copts` but we should also be adding `-Xcc -DSWIFT_PACKAGE` for the clang compilation step.

---

The following build file reproduces an issue, it shows a minimal Package.swift BUILD files after rules_swift_package_manager generates it:

```starlark
load("@build_bazel_rules_swift//swift:swift.bzl", "swift_library", "swift_library_group", "swift_binary")
load("@rules_swift_package_manager//swiftpkg/internal:generate_modulemap.bzl", "generate_modulemap")

swift_library_group(
  name = "Foo",
  visibility = ["//visibility:public"],
  deps = ["@swiftpkg_ios_svgkit//:SVGKit.rspm", "@swiftpkg_ios_svgkit//:SVGKit.rspm_modulemap"],
)

objc_library(
  name = "Foo.rspm",
  srcs = ["Foo.m"],
  enable_modules = True,
  module_name = "Foo",
  hdrs = ["Foo.h"],
  copts = [
    "-fblocks",
    "-fobjc-arc",
    "-fPIC",
    "-fmodule-name=Foo",
    "-DSWIFT_PACKAGE=1",
  ],
)

generate_modulemap(
  name = "Foo.rspm_modulemap",
  deps = [],
  hdrs = ["Foo.h"],
  module_name = "Foo",
  noop = False,
)

swift_library_group(
  name = "FooSwift",
  visibility = ["//visibility:public"],
  deps = [":FooSwift.rspm"],
)

swift_library(
  name = "FooSwift.rspm",
  tags = ["manual"],
  deps = [
    ":Foo.rspm",
    ":Foo.rspm_modulemap"
  ],
  copts = ["-DSWIFT_PACKAGE"],
  module_name = "FooSwift",
  srcs = ["Foo.swift"],
  always_include_developer_search_paths = True,
)

# Consumer example

swift_binary(
    name = "swift_package_copt_example",
    srcs = ["main.swift"],
    deps = [":FooSwift"],
)

```

With the following `Foo.h`

```c
#if SWIFT_PACKAGE
int foo(int x);
#else
int bar(int x); // should not be hit when building as a Swift package
#endif
```

When running:

```sh
bazel run swift_package_copt_example
```

The following error is shown:

```
Foo.swift:5:12: error: cannot find 'foo' in scope
    return foo(x)
           ^~~
```

This is because the `#if SWIFT_PACKAGE` case is not being hit. Fixing this requires updating the `copts` attribute of the `swift_library` to be: `copts = ["-DSWIFT_PACKAGE", "-Xcc", "-DSWIFT_PACKAGE"]`. 

Both are required because the first one is for the Swift compile and the second for the clang compile step.

--- a/swiftpkg/internal/bzl_selects.bzl
+++ b/swiftpkg/internal/bzl_selects.bzl
@@ -102,6 +102,11 @@ def _new_from_target_dependency_condition(kind, labels, condition = None):
         for p in spm_platforms.supported(condition.platforms)
     ]
 
+    # Because `spm_platforms.label` may transform two different platforms into
+    # the same Bazel label (e.g. `macos` and `driverkit`), we need to uniquify
+    # the conditions.
+    conditions = sets.to_list(sets.make(conditions))
+
     return [
         _new(kind = kind, value = labels, condition = c)
         for c in conditions
@@ -155,34 +160,33 @@ def _to_starlark(values, kind_handlers = {}, mutually_inclusive = False):
     # dict whose keys are the conditions and the value is the value for the
     # condition.
     selects_by_kind = {}
-    no_condition_results = sets.make()
+    no_condition_results = []
 
     for v in values:
         v_type = type(v)
         if v_type != "struct":
             if v_type == "list":
-                no_condition_results = sets.union(no_condition_results, sets.make(v))
+                no_condition_results.extend(v)
             else:
-                sets.insert(no_condition_results, v)
+                no_condition_results.append(v)
             continue
 
         # We are assuming that the select will always result in a list.
         # Hence, we wrap the transformed value in a list.
         kind_handler = kind_handlers.get(v.kind, _noop_kind_handler)
-        tvs_set = sets.make(lists.flatten(kind_handler.transform(v.value)))
+        tvs = lists.flatten(kind_handler.transform(v.value))
         if v.condition != None:
             # Collect all of the values associted with a condition.
             select_dict = selects_by_kind.get(v.kind, {})
-            condition_values = select_dict.get(v.condition, sets.make())
-            condition_values = sets.union(condition_values, tvs_set)
+            condition_values = select_dict.get(v.condition, []) + tvs
             select_dict[v.condition] = condition_values
             selects_by_kind[v.kind] = select_dict
         else:
-            no_condition_results = sets.union(no_condition_results, tvs_set)
+            no_condition_results = no_condition_results + tvs
 
     expr_members = []
-    if sets.length(no_condition_results) > 0:
-        expr_members.append(sets.to_list(no_condition_results))
+    if len(no_condition_results) > 0:
+        expr_members.append(no_condition_results)
     for (kind, select_dict) in selects_by_kind.items():
         kind_handler = kind_handlers.get(kind, _noop_kind_handler)
         sorted_keys = sorted(select_dict.keys())
@@ -191,13 +195,13 @@ def _to_starlark(values, kind_handlers = {}, mutually_inclusive = False):
             # Generate multiple select expressions for each condition.
             for k in sorted_keys:
                 new_dict = {
-                    k: sets.to_list(select_dict[k]),
+                    k: select_dict[k],
                 }
                 _append_select(expr_members, kind_handler, new_dict)
         else:
             # Combine all conditions of the same kind into one select expression.
             new_dict = {
-                k: sets.to_list(select_dict[k])
+                k: select_dict[k]
                 for k in sorted_keys
             }
             _append_select(expr_members, kind_handler, new_dict)

--- a/swiftpkg/internal/swiftpkg_build_files.bzl
+++ b/swiftpkg/internal/swiftpkg_build_files.bzl
@@ -106,6 +106,9 @@ def _swift_target_build_file(pkg_ctx, target):
         # SPM directive instructing the code to build as if a Swift package.
         # https://github.com/apple/swift-package-manager/blob/main/Documentation/Usage.md#packaging-legacy-code
         "-DSWIFT_PACKAGE",
+        # SPM directive instructing the code to build as if a Swift package for any clang modules.
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
     ]
 
     # GH046: Support plugins.

--- a/swiftpkg/tests/bzl_selects_tests.bzl
+++ b/swiftpkg/tests/bzl_selects_tests.bzl
@@ -158,11 +158,12 @@ def _to_starlark_test(ctx):
         struct(
             msg = "string values",
             khs = {},
-            vals = ["first", "second", "first"],
+            vals = ["-DFoo", "-Xcc", "-DFoo"],
             exp = """\
 [
-    "first",
-    "second",
+    "-DFoo",
+    "-Xcc",
+    "-DFoo",
 ]\
 """,
         ),
@@ -271,22 +272,27 @@ def _to_starlark_test(ctx):
             khs = {},
             vals = [
                 bzl_selects.new(
-                    value = "a",
+                    value = "-DFoo",
                     kind = "mykind",
                     condition = "//myconditions:alpha",
                 ),
                 bzl_selects.new(
-                    value = "b",
+                    value = "-DBar",
                     kind = "mykind",
                     condition = "//myconditions:beta",
                 ),
                 bzl_selects.new(
-                    value = "c",
+                    value = "-DZoo",
                     kind = "mykind",
                     condition = "//myconditions:alpha",
                 ),
                 bzl_selects.new(
-                    value = "a",
+                    value = "-Xcc",
+                    kind = "mykind",
+                    condition = "//myconditions:alpha",
+                ),
+                bzl_selects.new(
+                    value = "-DFoo",
                     kind = "mykind",
                     condition = "//myconditions:alpha",
                 ),
@@ -294,10 +300,12 @@ def _to_starlark_test(ctx):
             exp = """\
 select({
     "//myconditions:alpha": [
-        "a",
-        "c",
+        "-DFoo",
+        "-DZoo",
+        "-Xcc",
+        "-DFoo",
     ],
-    "//myconditions:beta": ["b"],
+    "//myconditions:beta": ["-DBar"],
     "//conditions:default": [],
 })\
 """,

--- a/swiftpkg/tests/swiftpkg_build_files_tests.bzl
+++ b/swiftpkg/tests/swiftpkg_build_files_tests.bzl
@@ -495,7 +495,11 @@ swift_library(
     name = "RegularSwiftTargetAsLibrary.rspm",
     always_include_developer_search_paths = True,
     alwayslink = True,
-    copts = ["-DSWIFT_PACKAGE"],
+    copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
+    ],
     module_name = "RegularSwiftTargetAsLibrary",
     package_name = "MyPackage",
     srcs = ["Source/RegularSwiftTargetAsLibrary/RegularSwiftTargetAsLibrary.swift"],
@@ -517,7 +521,11 @@ swift_library(
     name = "RegularTargetForExec.rspm",
     always_include_developer_search_paths = True,
     alwayslink = True,
-    copts = ["-DSWIFT_PACKAGE"],
+    copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
+    ],
     deps = ["@swiftpkg_mypackage//:RegularSwiftTargetAsLibrary.rspm"],
     module_name = "RegularTargetForExec",
     package_name = "MyPackage",
@@ -535,7 +543,11 @@ load("@build_bazel_rules_swift//swift:swift.bzl", "swift_test")
 
 swift_test(
     name = "RegularSwiftTargetAsLibraryTests.rspm",
-    copts = ["-DSWIFT_PACKAGE"],
+    copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
+    ],
     deps = ["@swiftpkg_mypackage//:RegularSwiftTargetAsLibrary.rspm"],
     module_name = "RegularSwiftTargetAsLibraryTests",
     package_name = "MyPackage",
@@ -553,6 +565,8 @@ load("@build_bazel_rules_swift//swift:swift.bzl", "swift_binary")
 swift_binary(
     name = "SwiftExecutableTarget.rspm",
     copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
         "-DSWIFT_PACKAGE",
         "-enable-experimental-feature",
         "BuiltinModule",
@@ -771,7 +785,11 @@ swift_library(
     name = "SwiftLibraryWithConditionalDep.rspm",
     always_include_developer_search_paths = True,
     alwayslink = True,
-    copts = ["-DSWIFT_PACKAGE"],
+    copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
+    ],
     deps = ["@swiftpkg_mypackage//:ClangLibrary.rspm"] + select({
         "@rules_swift_package_manager//config_settings/spm/platform:ios": ["@swiftpkg_mypackage//:RegularSwiftTargetAsLibrary.rspm"],
         "@rules_swift_package_manager//config_settings/spm/platform:tvos": ["@swiftpkg_mypackage//:RegularSwiftTargetAsLibrary.rspm"],
@@ -853,7 +871,11 @@ swift_library(
     name = "SwiftForObjcTarget.rspm",
     always_include_developer_search_paths = True,
     alwayslink = True,
-    copts = ["-DSWIFT_PACKAGE"],
+    copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
+    ],
     deps = ["@swiftpkg_mypackage//:ObjcLibraryDep.rspm"],
     features = ["swift.propagate_generated_module_map"],
     generates_header = True,
@@ -895,7 +917,11 @@ swift_library(
     name = "SwiftLibraryWithFilePathResource.rspm",
     always_include_developer_search_paths = True,
     alwayslink = True,
-    copts = ["-DSWIFT_PACKAGE"],
+    copts = [
+        "-DSWIFT_PACKAGE",
+        "-Xcc",
+        "-DSWIFT_PACKAGE",
+    ],
     data = [":SwiftLibraryWithFilePathResource.rspm_resource_bundle"],
     module_name = "SwiftLibraryWithFilePathResource",
     package_name = "MyPackage",

